meshmode/discretization/connection/modal.py

```diff
@@ -0,0 +1,373 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+import numpy.linalg as la
+import modepy as mp
+
+from meshmode.array_context import make_loopy_program
+from meshmode.dof_array import DOFArray
+from meshmode.discretization import (
+    NodalElementGroupBase,
+    ModalElementGroupBase,
+    InterpolatoryElementGroupBase)
+from meshmode.discretization.poly_element import QuadratureSimplexElementGroup
+from meshmode.discretization.connection.direct import DiscretizationConnection
+
+from pytools import memoize_in, keyed_memoize_in
+from pytools.obj_array import obj_array_vectorized_n_args
+
+
+class NodalToModalDiscretizationConnection(DiscretizationConnection):
+    """A concrete subclass of :class:`DiscretizationConnection`, which
+    maps nodal data to its modal representation. This connection can
+    be used with both unisolvent and non-unisolvent element groups.
+
+    .. note::
+
+        This connection requires that both nodal and modal discretizations
+        are defined on the *same* mesh.
+
+    .. attribute:: from_discr
+
+        an instance of :class:`meshmode.discretization.Discretization` containing
+        :class:`meshmode.discretization.NodalElementGroupBase` element groups
+
+    .. attribute:: to_discr
+
+        an instance of :class:`meshmode.discretization.Discretization` containing
+        :class:`meshmode.discretization.ModalElementGroupBase` element groups
+
+    .. attribute:: groups
+
+        a list of :class:`DiscretizationConnectionElementGroup`
+        instances, with a one-to-one correspondence to the groups in
+        :attr:`to_discr`.
+
+    .. automethod:: __call__
+
+    """
+
+    def __init__(self, from_discr, to_discr, allow_approximate_quad=False):
+        """
+        :arg from_discr: a :class:`meshmode.discretization.Discretization`
+            containing :class:`meshmode.discretization.NodalElementGroupBase`
+            element groups.
+        :arg to_discr: a :class:`meshmode.discretization.Discretization`
+            containing :class:`meshmode.discretization.ModalElementGroupBase`
+            element groups.
+        """
+
+        if not all([isinstance(grp, NodalElementGroupBase)
+                    for grp in from_discr.groups]):
+            raise ValueError("`from_discr` must be defined on nodal "
+                             "element groups to use this connection.")
+
+        if not all([isinstance(grp, ModalElementGroupBase)
+                    for grp in to_discr.groups]):
+            raise ValueError("`to_discr` must be defined on modal "
+                             "element groups to use this connection.")
+
+        if to_discr.mesh != from_discr.mesh:
```

Comment:
```suggestion
        if to_discr.mesh is not from_discr.mesh:
```
? Otherwise this calls [\_\_eq\_\_](https://github.com/inducer/meshmode/blob/ac37ba5bc53223f40c195e25b23778cf6a004329/meshmode/mesh/__init__.py#L906-L916). Which one was intended?

Reply:
I think calling `__eq__` is probably sufficient, but I like your suggestion as it's a stronger assertion. Will update :)